sklearn/model_selection/_validation.py

```diff
@@ -521,10 +523,34 @@ def _score(estimator, X_test, y_test, scorer, is_multimetric=False):
     return score
 
 
+class _MemoizedPredictEstimator:
+    def __init__(self, estimator):
+        self.estimator = estimator
+    def fit(self, X, y):
+        self.estimator.fit(X, y)
+    @if_delegate_has_method(delegate='estimator')
+    def predict(self, X):
+        if not hasattr(self, '_predictions'):
+            self._predictions = self.estimator.predict(X)
+        return self._predictions
+    @if_delegate_has_method(delegate='estimator')
+    def decision_function(self, X):
+        if not hasattr(self, '_decisions'):
+            self._decisions = self.estimator.decision_function(X)
+        return self._decisions
+
```

Comment:
No predict_proba or predict_log_proba? Are you these call decidion_function internally? Probabilistic non-linear predictors will be among those most benefiting from memoization, and I don't think they tend to implement decision_function.